File: python/cucumber_expressions/cucumber_expression_generator.py
import functools
import re

from cucumber_expressions.generated_expression import GeneratedExpression
from cucumber_expressions.parameter_type import ParameterType
from cucumber_expressions.parameter_type_matcher import ParameterTypeMatcher
from cucumber_expressions.combinatorial_generated_expression_factory import (
    CombinatorialGeneratedExpressionFactory,
)
from cucumber_expressions.parameter_type_registry import ParameterTypeRegistry


class CucumberExpressionGenerator:
    def __init__(self, parameter_type_registry: ParameterTypeRegistry):
        self.parameter_type_registry = parameter_type_registry

    def generate_expressions(self, text: str) -> list[GeneratedExpression]:
        parameter_type_combinations = []
        parameter_type_matchers = self.create_parameter_type_matchers(text)
        expression_template: str = ""
        pos = 0

        while True:
            matching_parameter_type_matchers: list[ParameterTypeMatcher] = []
            for parameter_type_matcher in parameter_type_matchers:
                advanced_parameter_type_matcher = parameter_type_matcher.advance_to(pos)
                if advanced_parameter_type_matcher.find:
                    matching_parameter_type_matchers.append(
                        advanced_parameter_type_matcher
                    )
            if matching_parameter_type_matchers:
                matching_parameter_type_matchers.sort(
                    key=functools.cmp_to_key(ParameterTypeMatcher.compare)
                )
                best_parameter_type_matcher = matching_parameter_type_matchers[0]
                best_parameter_type_matchers = list(
                    filter(
                        lambda m: ParameterTypeMatcher.compare(
                            m, best_parameter_type_matcher
                        )
                        == 0,
                        matching_parameter_type_matchers,
                    )
                )
                parameter_types = []
                for _parameter_type_matcher in best_parameter_type_matchers:
                    if _parameter_type_matcher.parameter_type not in parameter_types:
                        parameter_types.append(_parameter_type_matcher.parameter_type)
                parameter_types.sort(key=functools.cmp_to_key(ParameterType.compare))
                parameter_type_combinations.append(parameter_types)
                expression_template += self.escape(
                    text[pos : best_parameter_type_matcher.start]
                )
                expression_template += "{%s}"
                pos = best_parameter_type_matcher.start + len(
                    best_parameter_type_matcher.group
                )
            else:
                break

            if pos >= len(text):
                break
        expression_template += self.escape(text[pos:])
        return CombinatorialGeneratedExpressionFactory(
            expression_template, parameter_type_combinations
        ).generate_expressions()

    @staticmethod
    def escape(string: str) -> str:
        result = string.replace("%", "%%")
        result = result.replace(r"(", "\\(")
        result = result.replace(r"{", "\\{")
        return result.replace(r"/", "\\/")

    def create_parameter_type_matchers(self, text) -> list[ParameterTypeMatcher]:
        parameter_type_matchers = []
        for parameter_type in self.parameter_type_registry.parameter_types:
            if parameter_type.use_for_snippets:
                parameter_type_matchers += (
                    self.create_parameter_type_matchers_with_type(parameter_type, text)
                )
        return parameter_type_matchers

    @staticmethod
    def create_parameter_type_matchers_with_type(
        parameter_type, text
    ) -> list[ParameterTypeMatcher]:
        regexps = parameter_type.regexps
        return [
            ParameterTypeMatcher(parameter_type, re.compile(f"({regexp})"), text, 0)
            for regexp in regexps
        ]

File: elixir/lib/cucumber/cucumber_expressions/cucumber_expression_generator.ex
defmodule Cucumber.CucumberExpressions.CucumberExpressionGenerator do
  @moduledoc """
  Generates Cucumber Expression suggestions for a step text — the machinery
  behind "undefined step" snippets.
  """

  alias Cucumber.CucumberExpressions.CombinatorialGeneratedExpressionFactory
  alias Cucumber.CucumberExpressions.GeneratedExpression
  alias Cucumber.CucumberExpressions.ParameterType
  alias Cucumber.CucumberExpressions.ParameterTypeMatcher
  alias Cucumber.CucumberExpressions.ParameterTypeRegistry

  @doc """
  Generates suggested expressions for `text`, most specific first.
  """
  @spec generate_expressions(ParameterTypeRegistry.t(), String.t()) :: [GeneratedExpression.t()]
  def generate_expressions(%ParameterTypeRegistry{} = registry, text) do
    parameter_type_matchers = create_parameter_type_matchers(registry, text)

    {expression_template, parameter_type_combinations} =
      scan(parameter_type_matchers, text, 0, "", [])

    CombinatorialGeneratedExpressionFactory.generate_expressions(
      expression_template,
      parameter_type_combinations
    )
  end

  defp scan(parameter_type_matchers, text, pos, expression_template, parameter_type_combinations) do
    matching_parameter_type_matchers =
      parameter_type_matchers
      |> Enum.map(&ParameterTypeMatcher.advance_to(&1, pos))
      |> Enum.filter(&ParameterTypeMatcher.find?/1)

    case Enum.sort(matching_parameter_type_matchers, &ParameterTypeMatcher.sorts_before?/2) do
      [] ->
        finish(expression_template, text, pos, parameter_type_combinations)

      [best_parameter_type_matcher | _] = sorted ->
        # Collect the parameter types matching_parameter_type_matchers equally well, without
        # duplicates (a type with several regexps has several parameter_type_matchers).
        # Preferential types are listed first — users are most likely to want
        # those.
        parameter_types =
          sorted
          |> Enum.take_while(&ParameterTypeMatcher.same_rank?(&1, best_parameter_type_matcher))
          |> Enum.map(& &1.parameter_type)
          |> Enum.uniq()
          |> Enum.sort(&ParameterType.compare/2)

        parameter_type_combinations = [parameter_types | parameter_type_combinations]

        expression_template =
          expression_template <>
            escape(binary_part(text, pos, best_parameter_type_matcher.match.start - pos)) <>
            "{%s}"

        pos =
          best_parameter_type_matcher.match.start +
            byte_size(best_parameter_type_matcher.match.group)

        if pos >= byte_size(text) do
          finish(expression_template, text, pos, parameter_type_combinations)
        else
          scan(
            parameter_type_matchers,
            text,
            pos,
            expression_template,
            parameter_type_combinations
          )
        end
    end
  end

  defp finish(expression_template, text, pos, parameter_type_combinations) do
    tail =
      if pos >= byte_size(text) do
        ""
      else
        binary_part(text, pos, byte_size(text) - pos)
      end

    # `parameter_type_combinations` is accumulated by prepending, so reverse it back to
    # left-to-right expression_template order here.
    {expression_template <> escape(tail), Enum.reverse(parameter_type_combinations)}
  end

  defp create_parameter_type_matchers(registry, text) do
    for parameter_type <- ParameterTypeRegistry.parameter_types(registry),
        parameter_type.use_for_snippets,
        regexp <- parameter_type.regexps do
      ParameterTypeMatcher.new(
        parameter_type,
        Regex.compile!("(" <> regexp <> ")", "u"),
        text,
        0
      )
    end
  end

  defp escape(string) do
    string
    |> String.replace("%", "%%")
    |> String.replace("(", "\\(")
    |> String.replace("{", "\\{")
    |> String.replace("/", "\\/")
  end
end
